### PR TITLE
8324126: Error message for mistyping -XX:+Unlock...Options is not helpful

### PR DESCRIPTION
**Pull request by SoniaZaldana**
If we run the VM option to unlock experimental vm options without the preceding ‘+’, the error message is not very helpful: 

```Error: VM option 'UnlockExperimentalVMOptions' is experimental and must be enabled via -XX:+UnlockExperimentalVMOptions.```

With the proposed change, we should see a more descriptive error message, namely the lack of a +/- sign. 

```Missing +/- setting for VM option 'UnlockExperimentalVMOptions'```

Testing: runtime/commandLine regression tests pass. Also added a regression test to verify expected behaviour.

Edit (Jan 30): 
The expected output is both the unlock message and the well-formed error: 

```Error: VM option 'UnlockExperimentalVMOptions' is experimental and must be enabled via -XX:+UnlockExperimentalVMOptions.
Error: The unlock option must precede 'UnlockExperimentalVMOptions'.
Missing +/- setting for VM option 'UnlockExperimentalVMOptions'
Error: Could not create the Java Virtual Machine.
Error: A fatal exception has occurred. Program will exit.
```

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8324126](https://bugs.openjdk.org/browse/JDK-8324126): Error message for mistyping -XX:+Unlock...Options is not helpful (**Enhancement** - P4)

### Reviewers
 * [David Holmes](https://openjdk.org/census#dholmes) (@dholmes-ora - **Reviewer**) ⚠️ Review applies to [557f5cc6](https://git.openjdk.org/jdk/pull/17575/files/557f5cc6384e1f00da94eb19c392a88e9ecfe729)
 * [Thomas Stuefe](https://openjdk.org/census#stuefe) (@tstuefe - **Reviewer**) ⚠️ Review applies to [557f5cc6](https://git.openjdk.org/jdk/pull/17575/files/557f5cc6384e1f00da94eb19c392a88e9ecfe729)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/17575/head:pull/17575` \
`$ git checkout pull/17575`

Update a local copy of the PR: \
`$ git checkout pull/17575` \
`$ git pull https://git.openjdk.org/jdk.git pull/17575/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 17575`

View PR using the GUI difftool: \
`$ git pr show -t 17575`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/17575.diff">https://git.openjdk.org/jdk/pull/17575.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/17575#issuecomment-1911059158)

### CODE REVIEW

**Comment by tstuefe** [comment]
Mechanically, the fix looks okay. But we now will always, when missing +- for a locked option, print "missing +-":

Example:

"-XX:UseCompactObjectHeaders"
-> "missing +-"
User corrects:
"-XX:+UseCompactObjectHeaders"
-> "needs to be unlocked"

I think this is fine, and a minor concern, but let's ask @dholmes-ora.

The alternative would be to not make "UseDiagnosticVMOptions" option itself diagnostic, and vice versa with UseExperimentalVMOptions. Wonder what the reason for that was, @dholmes-ora ?

File: src/hotspot/share/runtime/arguments.cpp [hunk 1]
```diff
@@ -1,5 +1,5 @@
 /*
- * Copyright (c) 1997, 2023, Oracle and/or its affiliates. All rights reserved.
+ * Copyright (c) 1997, 2023, 2024, Oracle and/or its affiliates. All rights reserved.
```
Comment (dholmes-ora) [inline src/hotspot/share/runtime/arguments.cpp:2]: This copyright change is incorrect. Oracle copyrights only use 2 years: first and last, so 2023 simply becomes 2024.

**Comment by dholmes-ora** [comment]
> I think this is fine, and a minor concern, but let's ask @dholmes-ora.

So existing logic gives preference to whether the flag is unlocked, and then examines whether it is well-formed. The new code gives preference to whether it is well-formed and then whether it is unlocked. Can't you do both together?

> The alternative would be to not make "UseDiagnosticVMOptions" option itself diagnostic, and vice versa with UseExperimentalVMOptions. Wonder what the reason for that was, @dholmes-ora ?

I expect because it was considered appropriate. If those flags were themselves product flags then they would be subject to product flag rules.

File: test/hotspot/jtreg/runtime/CommandLine/TestVMOptions.java [hunk 2]
```diff
@@ -1,5 +1,5 @@
 /*
- * Copyright (c) 2014, 2023, Oracle and/or its affiliates. All rights reserved.
+ * Copyright (c) 2014, 2023, 2024, Oracle and/or its affiliates. All rights reserved.
```
Comment (dholmes-ora) [inline test/hotspot/jtreg/runtime/CommandLine/TestVMOptions.java:2]: Ditto re copyright.

**Review by dholmes-ora — CHANGES_REQUESTED** [review]

**Comment by dholmes-ora** [comment]
So now we will print the unlock message followed by any formatting error message? Can you provide sample output please. Thanks

**Review by dholmes-ora — APPROVED** [review]
Thanks. Looks good.

File: src/hotspot/share/runtime/arguments.cpp [hunk 3]
```diff
@@ -1120,27 +1120,26 @@ bool Arguments::process_argument(const char* arg,
...
-#ifdef PRODUCT
+    if (strlen(locked_message_buf) != 0) {
+      #ifdef PRODUCT
       bool mismatched = ((msg_type == JVMFlag::NOTPRODUCT_FLAG_BUT_PRODUCT_BUILD) ||
                          (msg_type == JVMFlag::DEVELOPER_FLAG_BUT_PRODUCT_BUILD));
       if (ignore_unrecognized && mismatched) {
         return true;
       }
-#endif
+      #endif
```
Comment (tstuefe) [inline src/hotspot/share/runtime/arguments.cpp:1130]: Please revert the indentation (above too) - we don't usually indent preprocessor directives .

File: src/hotspot/share/runtime/arguments.cpp [hunk 4]
```diff
@@ -1120,27 +1120,26 @@ bool Arguments::process_argument(const char* arg,
...
       jio_fprintf(defaultStream::error_stream(), "%s", locked_message_buf);
     }
+    if (found_flag->is_bool() && !has_plus_minus) {
+        jio_fprintf(defaultStream::error_stream(),
+          "Missing +/- setting for VM option '%s'\n", argname);
```
Comment (tstuefe) [inline src/hotspot/share/runtime/arguments.cpp:1135]: Fix indentation (2 spaces)

**Review by tstuefe — APPROVED** [review]
Small nits wrt indentation, otherwise good. If you fix them, I don't need another review - feel free to integrate.

> I expect because it was considered appropriate. If those flags were themselves product flags then they would be subject to product flag rules.

The more I think about this, the less sense this makes. UseExperimentalVMOptions is used to enable experimental VM options. That implies that it is a product flag itself, since you must be able to use in for a product VM that is not yet in ExperimentalVMOptions-allowed-state.

**Pull request closed without merging**